pkg/agent/discoverycache.go

```diff
@@ -24,264 +25,428 @@ const (
-		if primaryKey != "" {
-			existingAPIs[primaryKey] = true
-		} else {
-			existingAPIs[externalAPIID] = true
+		if err := dc.handleResource(svc); err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
-func (j *discoveryCache) fetchAPIServices() ([]*apiV1.ResourceInstance, error) {
+func (dc *discoveryCache) handleServiceInstance() error {
+	dc.logger.WithField("kind", "APIServiceInstance").Trace("fetching API Service Instances and updating cache")
+	dc.instanceCacheLock.Lock()
+	defer dc.instanceCacheLock.Unlock()
+
 	query := map[string]string{
 		apic.FieldsKey: apiServerFields,
 	}
 
-	return GetCentralClient().GetAPIV1ResourceInstancesWithPageSize(
-		query, agent.cfg.GetServicesURL(), apiServerPageSize,
+	svcInstanceLink := mv1.NewAPIServiceInstance("", dc.envName).GetKindLink()
+	url := dc.formatResourceURL(svcInstanceLink)
+
+	serviceInstances, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(
+		query, url, apiServerPageSize,
 	)
+	if err != nil {
+		e := utilErrors.Wrap(ErrUnableToGetAPIV1Resources, err.Error()).FormatError("APIServiceInstances")
+		dc.logger.Error(e)
+		return e
+	}
+
+	for _, instance := range serviceInstances {
+		if err := dc.handleResource(instance); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (j *discoveryCache) updateAPIServiceInstancesCache() {
-	j.instanceCacheLock.Lock()
-	defer j.instanceCacheLock.Unlock()
+func (dc *discoveryCache) handleCategories() error {
+	dc.logger.WithField("kind", "Category").Trace("fetching Categories and updating cache")
 
+	// Update cache with published resources
 	query := map[string]string{
 		apic.FieldsKey: apiServerFields,
 	}
 
-	serviceInstances, err := GetCentralClient().GetAPIV1ResourceInstancesWithPageSize(
-		query, agent.cfg.GetInstancesURL(), apiServerPageSize,
+	categoriesLink := catalog.NewCategory("").GetKindLink()
+	url := dc.formatResourceURL(categoriesLink)
+
+	categories, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(
+		query, url, apiServerPageSize,
 	)
 	if err != nil {
-		j.logger.Error(utilErrors.Wrap(ErrUnableToGetAPIV1Resources, err.Error()).FormatError("APIServiceInstances"))
-		return
+		return err
 	}
 
-	for _, instance := range serviceInstances {
-		id, _ := util.GetAgentDetailsValue(instance, defs.AttrExternalAPIID)
-		if id == "" {
-			continue // skip instance without external api id
+	for _, category := range categories {
+		if err := dc.handleResource(category); err != nil {
+			return err
 		}
-		agent.cacheManager.AddAPIServiceInstance(instance)
 	}
+	return nil
 }
 
-func (j *discoveryCache) updateCategoryCache() {
-	j.logger.Trace("updating category cache")
+func (dc *discoveryCache) handleAccessControlList() error {
+	dc.logger.WithField("kind", "AccessControlList").Trace("fetching AccessControlList and updating cache")
 
 	// Update cache with published resources
-	existingCategories := make(map[string]bool)
 	query := map[string]string{
 		apic.FieldsKey: apiServerFields,
```

Comment:
@jcollins-axway Are these the right fields to query for on an access control list?

Reply:
@jcollins-axway Can we remove the query for all the resources we fetch then?